### PR TITLE
[PRISM] Compile RescueNode

### PR DESCRIPTION
**Pull request by eightbitraptor**

### CODE REVIEW

**Review by kddnewton — APPROVED** [review]
Looks great!

File: prism_compile.c [hunk 1]
```diff
@@ -1852,6 +1857,63 @@ pm_compile_node(rb_iseq_t *iseq, const pm_node_t *node, LINK_ANCHOR *const ret,
...
+                                                                  ISEQ_BODY(iseq)->location.label),
+                                                    ISEQ_TYPE_RESCUE, 1);
+            lstart->rescued = LABEL_RESCUE_BEG;
+            lend->rescued = LABEL_RESCUE_END;
+            ADD_LABEL(ret, lstart);
+
+            bool prev_in_rescue = ISEQ_COMPILE_DATA(iseq)->in_rescue;
+            ISEQ_COMPILE_DATA(iseq)->in_rescue = true;
+            {
+                pm_node_list_t node_list = begin_node->statements->body;
```
Comment (jemmaissroff) [inline prism_compile.c:1876]: Is it possible that the begin_node has no statements?

File: prism_compile.c [hunk 2]
```diff
@@ -1852,6 +1857,63 @@ pm_compile_node(rb_iseq_t *iseq, const pm_node_t *node, LINK_ANCHOR *const ret,
...
+                if (node_list.size > 0) {
```
Comment (jemmaissroff) [inline prism_compile.c:1877]: I think this if / else can be simplified to `PM_COMPILE_NOT_POPPED(begin_node->statements)`

File: prism_compile.c [hunk 3]
```diff
@@ -1852,6 +1857,63 @@ pm_compile_node(rb_iseq_t *iseq, const pm_node_t *node, LINK_ANCHOR *const ret,
...
+            if (begin_node->else_clause) {
+                PM_POP_UNLESS_POPPED;
+                PM_COMPILE((pm_node_t *)begin_node->else_clause);
+            }
+
+            ADD_LABEL(ret, lend);
+            ADD_INSN(ret, &dummy_line_node, nop);
+            ADD_LABEL(ret, lcont);
+
+            if (begin_node->else_clause) {
```
Comment (jemmaissroff) [inline prism_compile.c:1898]: If we're calling `PM_POP_IF_POPPED` in either case, do we need this if / else block?

File: prism_compile.c [hunk 4]
```diff
@@ -3572,6 +3633,50 @@ pm_compile_node(rb_iseq_t *iseq, const pm_node_t *node, LINK_ANCHOR *const ret,
...
+
+        pm_node_list_t exception_list = rescue_node->exceptions;
+        if (exception_list.size > 0) {
+            for (size_t i = 0; i < exception_list.size; i++) {
+                ADD_GETLOCAL(ret, &dummy_line_node, LVAR_ERRINFO, 0);
+                PM_COMPILE(exception_list.nodes[i]);
+                ADD_INSN1(ret, &dummy_line_node, checkmatch, INT2FIX(VM_CHECKMATCH_TYPE_RESCUE));
+                ADD_INSN1(ret, &dummy_line_node, branchif, excep_match);
+            }
+            ADD_INSN1(ret, &dummy_line_node, jump, rescue_end);
```
Comment (jemmaissroff) [inline prism_compile.c:3651]: Looks like this is happening in either case, can we pull it out of the if / else?

File: test/ruby/test_compile_prism.rb [hunk 5]
```diff
@@ -769,6 +769,69 @@ def test_RedoNode
...
+      assert_prism_eval(<<-CODE)
+        a = 1
+        begin
+          b = 2
+          raise "bang"
+        rescue
+          c = 3
+        end
+        a + b + c
+      CODE
```
Comment (jemmaissroff) [inline test/ruby/test_compile_prism.rb:832]: Can we add a test case with an empty begin?

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: prism_compile.c [hunk 6]
```diff
@@ -1852,6 +1857,55 @@ pm_compile_node(rb_iseq_t *iseq, const pm_node_t *node, LINK_ANCHOR *const ret,
...
+                                                    rb_str_concat(rb_str_new2("rescue in"),
+                                                                  ISEQ_BODY(iseq)->location.label),
+                                                    ISEQ_TYPE_RESCUE, 1);
+            lstart->rescued = LABEL_RESCUE_BEG;
+            lend->rescued = LABEL_RESCUE_END;
+            ADD_LABEL(ret, lstart);
+
+            bool prev_in_rescue = ISEQ_COMPILE_DATA(iseq)->in_rescue;
+            ISEQ_COMPILE_DATA(iseq)->in_rescue = true;
+            {
```
Comment (jemmaissroff) [inline prism_compile.c:1875]: Nit, but what are these braces for?

**Review by jemmaissroff — APPROVED** [review]
One nit, otherwise lgtm, thanks!
